File: src/main/java/me/itdog/rssthis/evaluate/XPathEvaluator.java
package me.itdog.rssthis.evaluate;

import net.sf.saxon.s9api.*;
import org.apache.commons.io.IOUtils;

import javax.xml.transform.Source;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.nio.charset.StandardCharsets;

public abstract class XPathEvaluator {

    Processor processor;
    XPathCompiler compiler;
    Source source;
    XdmNode root;

    public XPathEvaluator() {
        processor = new Processor(false);
        compiler = processor.newXPathCompiler();
    }

    abstract Source buildSource(InputStream is) throws SaxonApiException;

    public void setSource(URI uri) throws IOException, SaxonApiException {
        InputStream is = uri.toURL().openConnection().getInputStream();
        source = buildSource(is);
        root = processor.newDocumentBuilder().build(source);
    }

    public void setSource(File file) throws IOException, SaxonApiException {
        source = buildSource(new FileInputStream(file));
        root = processor.newDocumentBuilder().build(source);
    }

    public void setSource(String src) throws SaxonApiException {
        this.source = buildSource(IOUtils.toInputStream(src, StandardCharsets.UTF_8));
        root = processor.newDocumentBuilder().build(source);
    }

    public XdmValue evaluate(String xpath) throws SaxonApiException {
        return compiler.evaluate(xpath, root);
    }
}

File: src/main/java/me/itdog/rssthis/evaluate/XPathEvaluator.kt
package me.itdog.rssthis.evaluate

import net.sf.saxon.s9api.Processor
import net.sf.saxon.s9api.XdmNode
import net.sf.saxon.s9api.XdmValue
import java.io.File
import java.io.InputStream
import java.net.URI
import java.nio.charset.StandardCharsets
import javax.xml.transform.Source


abstract class XPathEvaluator {

    protected val processor = Processor(false)
    protected val compiler = processor.newXPathCompiler()
    protected lateinit var source: Source
    protected lateinit var root: XdmNode

    abstract fun buildSource(inputStream: InputStream): Source

    fun setSource(uri: URI) {
        source = buildSource(uri.toURL().openStream())
        root = processor.newDocumentBuilder().build(source)
    }

    fun setSource(file: File) {
        source = buildSource(file.inputStream())
        root = processor.newDocumentBuilder().build(source)
    }

    fun setSource(string: String) {
        source = buildSource(string.byteInputStream(StandardCharsets.UTF_8))
        root = processor.newDocumentBuilder().build(source)
    }

    fun evaluate(expression: String): XdmValue {
        return compiler.evaluate(expression, root)
    }
}
